docs/proposals/20230123-azure-service-operator.md

```diff
@@ -0,0 +1,214 @@
+
+Also, use of the `clusterctl move` command will require extra manual steps to move ASO resources as documented here: https://azure.github.io/azure-service-operator/introduction/frequently-asked-questions/#what-is-the-best-practice-for-transferring-aso-resources-from-one-cluster-to-another.
+
+### Security Model
+
+ASO requires credentials to interact with Azure which are documented here: https://azure.github.io/azure-service-operator/introduction/authentication/. Specifically, ASO needs the values of `$AZURE_CLIENT_ID`, `$AZURE_CLIENT_SECRET`, `$AZURE_SUBSCRIPTION_ID`, and `$AZURE_TENANT_ID` for Service Principal authentication. These credentials must be defined in a Kubernetes Secret that either applies to all ASO resources in all namespaces, all ASO resources in a particular namespace, or individual ASO resources referring to the Secret in its `serviceoperator.azure.com/credential-from` annotation. ASO allows different credentials to be used for different resources to enable multitenancy, similar to CAPZ and its AzureClusterIdentity. To mediate between the two mechanisms, CAPZ will ensure a cluster's AzureClusterIdentity or default parameters from environment variables are reflected in a Secret to be used by ASO and ensure each ASO resource it manages is created with the annotation pointing to the Secret.
+
+Consumers of ASO's APIs (such as CAPZ) also need Kubernetes RBAC permissions to interact with ASO resources. For example, this ClusterRole snippet is required by CAPZ to manage resource groups with ASO:
+```yaml
+- apiGroups:
+  - resources.azure.com
+  resources:
+  - resourcegroups
+  - resourcegroups/status
+  verbs:
+  - get
+  - list
+  - watch
+  - create
+  - update
+  - delete
+```
+The above permissions may overlap with ASO resources already installed by a user not intended to be managed by CAPZ.
+
+### Risks and Mitigations
+
+The most immediate risks with adopting ASO are regressions of various forms. Firstly, Azure API calls are core to CAPZ's functionality, so even small behavioral differences may cause issues preventing Azure resources from being reconciled as expected. Next, the modified reconciliation flow may cause operations against workload cluster to take longer to complete, in particular if it takes more reconciliations to perform a task. Finally, integration with ASO may introduce new or worsen existing bottlenecks limiting how many CAPZ resources can be managed. Scaling bottlenecks might be Azure API rate limits or management cluster Kubernetes API server load because of the increased number of Kubernetes resources.
```

Comment:
Let's add some mitigations for these risks: for example, how can we leverage unit testing and existing e2e tests to prevent behavioral regressions? Are there any metrics or tests we can add to measure performance and scaling? what does a "roll back" look like if we find a regression?

Reply:
should have kept reading... https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/3113/files#diff-83dfaf571cf340f67871425a334a6b019c7715c0ef5b1adae249017d59cada46R193 #2123 

maybe link to the test plan section in here?

Reply:
Yep, I'll link that section here.

> what does a "roll back" look like if we find a regression?

Is there something more to consider here than "PR to revert the change" like we would normally do?

Reply:
> Is there something more to consider here than "PR to revert the change" like we would normally do?

no that's basically it, but worth mentioning that it is that simple (and should be backwards compatible).